src/swarm-engine/Swarm/Game/Step/Util/Command.hs

```diff
@@ -150,18 +151,42 @@ onTarget rid act = do
               then deleteRobot rid
               else robotMap . ix rid .= tgt'
 
+grantAchievementForRobot ::
+  (HasRobotStepState sig m, Has (Lift IO) sig m) =>
+  GameplayAchievement ->
+  m ()
+grantAchievementForRobot a = do
+  sys <- use systemRobot
+  let isValidRobotType = not sys || robotTypeRequired == ValidForSystemRobot
+  when isValidRobotType $
+    grantAchievement a
+ where
+  ValidityConditions robotTypeRequired _ = getValidityRequirements a
+
+checkGameModeAchievementValidity ::
+  Has (State GameState) sig m =>
+  GameplayAchievement ->
+  m Bool
+checkGameModeAchievementValidity a = do
+  creative <- use creativeMode
+  return $ not creative || gameplayModeRequired == ValidInCreativeMode
+ where
+  ValidityConditions _ gameplayModeRequired = getValidityRequirements a
+
 grantAchievement ::
```

Comment:
Can you add a comment to `grantAchievement` explaining that one probably wants to call `grantAchievementForRobot` instead of calling `grantAchievement` directly?

Reply:
Added in #1785.